modules/chapter1/pages/section2.adoc

```diff
@@ -1,3 +1,355 @@
+****
+
+The entire pipeline is here link:sample-pipeline-full.py[Pipeline]
+
+We're not going to go through all of it but focus on the key aspects of it. 
+
+The actual pipeline is defined by the following function
+
+****
+[source,python]
+include::sample-pipeline-full.py[lines=450..454]
+****
+
+The '@dsl.pipeline' parameters provide the name and description if you were uploading the pipeline via an API call. The DSP UI overwrites these values.
+
+The function *_edgetest_pipeline_* is the implementation of the pipeline.
+
+=== Pipeline Parameters
+The pipeline has four parameters
+
+* _file_obj_ and _src_bucket_ refer to S3 bucket details and can be ignored.
+* _VIN_ is the edge device identifier and has a default value of 412356
+* _epoch_count_ is the number of training epochs to be used
+
+In the _Create Run_ UI these parameters are available so that users can override the values as they need
+
+****
+image::pipeline-parameters.png[]
+****
+
+=== Pipeline Steps
+The file contains the following python functions which roughly correspond to the steps in the diagram above
+
+* load_trigger_data()
+* prep_data_train_model()
+* model_upload_notify()
+* model_inference()
+
+These functions are mapped into individual containers by using the _create_component_from_func_ function. You can specify the container _base_image_ to use as well as the package to be installed into the container.
+
+****
+[source,python]
+include::sample-pipeline-full.py[lines=433..443]
+****
+
+The python functions can be used in multiple containers; in the example the _prep_data_train_model_ function is wrapper in the _prep_data_train_op_ and the _prep_inference_data_op_ containers.
+
+The pipeline execution _graph_ is created using the following code
+
+****
+[source,python]
+include::sample-pipeline-full.py[lines=471]
+
+[source,python]
+include::sample-pipeline-full.py[lines=477]
+
+[source,python]
+include::sample-pipeline-full.py[lines=482]
+
+[source,python]
+include::sample-pipeline-full.py[lines=492]
+
+[source,python]
+include::sample-pipeline-full.py[lines=495]
+
+****
+
+The execution order of the graph is top down but also can be controlled by using the *_.after()_* operator. The following diagram shows the order of execution.
+
+****
+image::pipeline-graph.png[]
+****
+
+Also visable in the OpenShift AI user interface
+
+****
+image::oai-pipeline-graph.png[]
+****
+
+=== Pipeline Parameter Passing
+As each step of our pipeline is an independent container, passing values between these steps requires a persistent store to be used between steps. This is where the associated S3 bucket that we configured earlier comes into play.
+
+In our pipeline there is a number of data values passed between the steps. For example :
+
+* The output of the _trigger_data_ step is passed into the input of the _prep_train_data_op_ step.
+* The _upload_model_op_ uses the output of the _trigger_data_ and one of the *_named outputs_* of the _prep_train_data_ step. 
+
+==== Difference 
```

Comment:
Are we missing content in this `Difference` section?

Reply:
yep not finished as yet